blazar/tests/enforcement/test_enforcement.py

```diff
@@ -0,0 +1,209 @@
+
+    def test_format_context(self):
+
+        formatted_context = self.enforcement.format_context(context.current())
+
+        expected = dict(user_id=111, project_id=222, region_name=self.region,
+                        auth_url='https://fakeauth.com')
+
+        self.assertDictEqual(expected, formatted_context)
+
+    def test_check_create(self):
+        allocation_candidates = {'virtual:instance': [get_fake_host('1')]}
+        lease_values = get_fake_lease()
+        reservations = list(lease_values['reservations'])
+
+        del lease_values['reservations']
+        ctx = context.current()
+
+        check_create = self.patch(self.enforcement.enabled_filters[0],
+                                  'check_create')
+
+        self.enforcement.check_create(ctx, lease_values, reservations,
+                                      allocation_candidates)
+
+        formatted_lease = self.enforcement.format_lease(
+            lease_values, reservations, allocation_candidates)
+        formatted_context = self.enforcement.format_context(ctx)
+
+        check_create.assert_called_once_with(formatted_context,
+                                             formatted_lease)
+
+    def test_check_update(self):
+        current_allocations = {'virtual:instance': [get_fake_host('1')]}
+        lease = get_fake_lease()
+        reservations = list(lease['reservations'])
+
+        new_lease_values = get_fake_lease(end_date=u'2014-02-07 13:37')
+        new_reservations = list(new_lease_values['reservations'])
+        allocation_candidates = {'virtual:instance': [get_fake_host('2')]}
+
+        del new_lease_values['reservations']
+        ctx = context.current()
+
+        check_update = self.patch(self.enforcement.enabled_filters[0],
+                                  'check_update')
+
+        self.enforcement.check_update(
+            ctx, lease, new_lease_values, current_allocations,
+            allocation_candidates, reservations, new_reservations)
+
+        formatted_context = self.enforcement.format_context(ctx)
+        formatted_lease = self.enforcement.format_lease(
+            lease, reservations, current_allocations)
+        new_formatted_lease = self.enforcement.format_lease(
+            new_lease_values, new_reservations, allocation_candidates)
+
+        check_update.assert_called_once_with(
+            formatted_context, formatted_lease, new_formatted_lease)
+
+    def test_on_end(self):
+        allocations = {'virtual:instance': [get_fake_host('1')]}
+        lease = get_fake_lease()
+        ctx = context.current()
+
+        on_end = self.patch(self.enforcement.enabled_filters[0], 'on_end')
+
+        self.enforcement.on_end(ctx, lease, allocations)
+
+        formatted_context = self.enforcement.format_context(ctx)
+        formatted_lease = self.enforcement.format_lease(
+            lease, lease['reservations'], allocations)
+
+        on_end.assert_called_once_with(formatted_context, formatted_lease)
```

Comment:
Similarly, ensure that if on_end fails for some reason, nothing bad happens.

Reply:
So I added a test that will raise an exception and verify that the lease will not delete is on_end fails. However, maybe we don't want that behavior. In which case, we need to add on_end to a try/except block and then ensure the lease is still deleted provided no other errors.

Reply:
I think we want to avoid failing the lease delete if on_end fails. It should log an error in the logs, but otherwise the lease should proceed to delete OK.

Reply:
Updated. Applied try/except block to delete_lease and end_lease.